### PR TITLE
Add Elixir/Hex package

### PR DESCRIPTION
**Pull request by nihalgonsalves**
Back again with another package, this time for Elixir! Let me know what you think :)

P.s. if you create a https://hex.pm/ user for Razorpay I can add it as an owner. 

Todo:

- [ ] wercker task

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -0,0 +1,77 @@
+defmodule IFSC.Mixfile do
+  use Mix.Project
+
+  def project do
+    [
+      app: :ifsc,
+      version: "0.1.0",
```
Comment (captn3m0) [inline mix.exs]: Keep version same as the release version (`1.0.14`)

File: mix.lock [hunk 2]
```diff
@@ -0,0 +1,19 @@
+%{
```
Comment (captn3m0) [inline mix.lock]: TIL mix.lock is committed even for packages.

File: src/elixir/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+## Development and Publishing
+
+Remember to run `mix ifsc.copy_json` to populate the `priv/ifsc-data` directory,
+or the `Razorpay.IFSC.validate` function will not run.
+
+This is because the root repository generates the IFSC data in `src/`, and to
+access these files when the `ifsc` package is included in another project,
+they must be inside `priv/`.
+
+Thus `priv/ifsc-data` is gitignore'd but populated and included when publishing to Hex.
```
Comment (captn3m0) [inline src/elixir/README.md]: Can you just symlink the directory instead?

**Comment by captn3m0** [comment]
@nihalgonsalves Thanks a lot for this 👍 

I'll notify once I create the hex.pm user for razorpay.

Meanwhile, just a heads-up, we'll be making changes to all our client SDKs to accommodate for https://github.com/razorpay/ifsc/issues/57.

**Comment by nihalgonsalves** [comment]
Thanks, @captn3m0 

1. Once you have a Hex.pm user, I think its better you do the version change and try out the publishing process yourself.
2. `mix.lock` removed
3. I did experiment with using symlinks but they didn't seem to work, perhaps due to Elixir's build process. I don't think the current process is too bad: the usual `hex.publish` process will automatically generate/overwrite these before publishing (due to the alias I added).

Regarding the client redirect, I'm not sure if that's the best idea. Though it isn't disallowed by the spec, providing a body in a 302 is not really standard practice. May I suggest API versioning, or having a client-specified parameter that allows opt-in to redirection? (e.g. even something simple like `GET https://ifsc.razorpay.com/KKBK0000261?allow_ifsc_redirect=true`)

This way clients that may be expecting a 404 to validate an IFSC code (if someone uses the packages in online-only mode) will still receive a 404, and newer clients can account for the redirects/newly assigned codes.

Also, if other users are directly calling the API, the behaviour may be undefined - depending on what HTTP client and settings they are using.

**Comment by captn3m0** [comment]
The docs say mix.lock should be committed (looks like it is true for libraries as well, but you'll be a better judge): https://mandarvaze.bitbucket.io/posts/what-is-mixlock-file.html

I liked the idea of using 302+body, just because it is allowed by the spec (apache default redirect includes a body with a meta refresh for eg). I don't wanna do 404 +redirects, since that would be breaking HTTP right now. I'll keep the feedback in mind though, and this is not final, just a heads-up.

**Comment by nihalgonsalves** [comment]
Whether we commit `mix.lock` or not to the repository only affects the development environment. Its definitely committed for normal projects (like a Gemfile.lock) but I'm not sure what the consensus for packages is. I've seen repos both ways.

What definitely needed to be removed was `mix.lock` from the package files list. I think I'll add it back to the repo (and of course leave it out of the files list).

Can you put the wercker task in? Is there anything else that needs to be addressed before merging?

**Comment by captn3m0** [comment]
Just wercker, release and then merge.

I'll also add you to the repo collaborators so you can commit directly.

**Comment by captn3m0** [comment]
Setting up wercker will require lots of small commits (from past experience). I'll test this locally, merge, and take up wercker in a separate PR.

**Comment by nihalgonsalves** [comment]
@captn3m0 sounds good!

**Pull request merged**